### PR TITLE
Fix #825 - building agora based on agora-builder image as opposed to alpine image

### LINKED ISSUE #825: Provide a "base docker build" image
Currently, building the docker image from scratch can be quite slow, as it installs 45 packages, including an upgraded GCC.
Since we're depending on edge, this happens quite often, slowing developers and CI alike.

### PR DESCRIPTION
**Pull request by ferencdg**
building agora based on agora-builder image as opposed to alpine image

### CODE REVIEW

**Review by Geod24 — APPROVED** [review]

**Review by TrustHenry — APPROVED** [review]
LGTM

**Comment by AndrejMitrovic** [comment]
If you rebase it should go green.

**Comment by Geod24** [comment]
Looks like you added an extra commit here.

File: Dockerfile [hunk 1]
```diff
@@ -1,7 +1,6 @@
 # Build Agora from source
```
Comment (ferencdg) [inline Dockerfile:1]: this issue of #825 is very trivial, so I included my #1024 changes here too, please let me know if you want me to separate the two

File: source/agora/network/NetworkManager.d [hunk 2]
```diff
@@ -326,10 +328,10 @@ public class NetworkManager
     protected ConnectionTask[Address] connection_tasks;
 
     /// List of validator clients
-    protected NetworkClient[] validators;
+    protected DList!NetworkClient validators;
```
Comment (ferencdg) [inline source/agora/network/NetworkManager.d]: cannot use single list here, as addition to the end of SList is O(n)

File: source/agora/network/NetworkManager.d [hunk 3]
```diff
@@ -660,21 +662,21 @@ public class NetworkManager
     }
 
     ///
-    private bool minPeersConnected ()  pure nothrow @safe @nogc
+    private bool minPeersConnected ()  pure nothrow @safe
     {
         return this.required_peer_keys.length == 0 &&
-            this.peers.length >= this.node_config.min_listeners;
+            this.peers[].walkLength >= this.node_config.min_listeners;
```
Comment (ferencdg) [inline source/agora/network/NetworkManager.d]: This is an O(n) operation on DList, but in case the peer list is not long, it should be fine. Please let me know if I should find an other DList implementation that can produce the list size in O(1)

File: source/agora/network/NetworkManager.d [hunk 4]
```diff
@@ -660,21 +662,21 @@ public class NetworkManager
     }
 
     ///
-    private bool minPeersConnected ()  pure nothrow @safe @nogc
+    private bool minPeersConnected ()  pure nothrow @safe
```
Comment (ferencdg) [inline source/agora/network/NetworkManager.d]: walkLength needs GC

File: source/agora/network/NetworkManager.d [hunk 5]
```diff
@@ -660,21 +662,21 @@ public class NetworkManager
...
     {
         return this.required_peer_keys.length == 0 &&
-            this.peers.filter!(node =>
+            this.peers[].filter!(node =>
                 !this.banman.isBanned(node.address)).count >= this.node_config.max_listeners;
     }
 
     /// Returns: the list of node IPs this node is connected to
-    public NodeInfo getNetworkInfo () pure nothrow @safe @nogc
+    public NodeInfo getNetworkInfo () pure nothrow @safe
```
Comment (linked0) [inline source/agora/network/NetworkManager.d:679]: This kind of modification should be another PR or commit because it's not related to this commit.

File: source/agora/node/FullNode.d [hunk 6]
```diff
@@ -238,7 +238,7 @@ public class FullNode : API
     }
 
     /// GET: /node_info
-    public override NodeInfo getNodeInfo () pure nothrow @safe @nogc
+    public override NodeInfo getNodeInfo () pure nothrow @safe
```
Comment (linked0) [inline source/agora/node/FullNode.d:241]: The same as the above comment.

**Review by linked0 — COMMENTED** [review]
The second and third commits are not related to this PR because the title of this PR doesn't explain the commits it contains. So you should separate the commits into another PR.
If any questions, please let us know on the Slack.
Thanks.

**Review by ferencdg — COMMENTED** [review]
> The second and third commits are not related to this PR because the title of this PR doesn't explain the commits it contains. So you should separate the commits into another PR.
> If any questions, please let us know on the Slack.
> Thanks.

yes, already removed and will create a separate PR

**Comment by AndrejMitrovic** [comment]
It should be fairly easy to wrap `DList` in another struct which keeps track of the length. Something like:

```D
struct List (T)
{
    private DList!T dlist;
    private size_t _length;

    public size_t length () @nogc { return _length; }
    // implement forwarding of opSlice / insertBack, update _length when insertBack is called
}
```

**Comment by ferencdg** [comment]
> It should be fairly easy to wrap `DList` in another struct which keeps track of the length. Something like:
> 
> ```d
> struct List (T)
> {
>     private DList!T dlist;
>     private size_t _length;
> 
>     public size_t length () @nogc { return _length; }
>     // implement forwarding of opSlice / insertBack, update _length when insertBack is called
> }
> ```

I was also thinking about this idea, but there aren't many elements in the container. Discussed it with Mathias, and for now we would just use the original DList.

File: Dockerfile (hunk 1 above)
Comment (linked0) [reply Dockerfile:1]: I know what you intended as you explained. But, please refer to [the example of a simple PR](https://github.com/bpfkorea/agora/pull/130)

In addition, it would be good to add the issue number in the pull request's description like this.
`Related to #825` or `Fixes #825`
Please refer to other PRs.

Thanks!

**Comment by linked0** [comment]
I'm removing improper comments about function attributes like @nogc.
Sorry for the confusion.

**Comment by Geod24** [comment]
Can you clean up the commits ?
